### DISCUSSION TITLE
Which Script for TASMOTA Smart Meter Interface with efr SGM-C8

### DISCUSSION THREAD

**Discussion by Amirosh1** [arendst/Tasmota#21415] [Support request]
I am looking for a Script for my Meter efr SGM-C8  not  C4  please help me . thanks

**Comment by perry010** [comment]
>D
>B
->sensor53 r
>M 1
+1,3,s,16,9600,ENERGY
1,77070100100700ff@1,Aktuelle Leistung,W,Power,0
1,77070100010800ff@1000,Bezug,kWh,Total,4
1,77070100020800ff@1000,Einspeisung,kWh,Supply,4
1,77070100200700ff@1,Spannung L1,V,Voltage,1
1,77070100340700ff@1,Spannung L2,V,Voltage_L2,1
1,77070100480700ff@1,Spannung L3,V,Voltage_L3,1
1,770701001f0700ff@1,Strom L1,A,Current,2
1,77070100330700ff@1,Strom L2,A,Current_L2,2
1,77070100470700ff@1,Strom L3,A,Current_L3,2
1,77070100240700FF@1,Wirkleistung L1,W,wirkleistung_l1,0
1,77070100380700FF@1,Wirkleistung L2,W,wirkleistung_l2,0
1,770701004C0700FF@1,Wirkleistung L3,W,wirkleistung_l3,0
1,77070100510701ff@1,Phasenabweichung L2-L1,°,phase_angle_L2_L1,0
1,77070100510702ff@1,Phasenabweichung L3-L1,°,phase_angle_L3_L1,0
1,77070100510704ff@1,Phasenabweichung I/U L1,°,phase_angle_L1,1 
1,7707010051070fff@1,Phasenabweichung I/U L2,°,phase_angle_L2,1  
1,7707010051071aff@1,Phasenabweichung I/U L3,°,phase_angle_L3,1 
1,770701000e0700ff@1,Frequenz,Hz,Freq,1
1,77070100600100ff@#,Server-ID,,ID,0
#

**Reply by Sympatron** [reply]
This looks almost exactly like the SGM-C4 script. Does that mean that they use the same codes?

**Reply by FaserF** [reply]
Yes, just tested it on my C8, the C4 Script works indeed.

**Reply by Smitty367** [reply]
Hello,

I also got the C8 today and wanna read it by tasmota. I another posting I read that for C8 the line

+1,3,s,16,9600,ENERGY

has to be repaced by

+1,5,s,16,9600,ENERGY     (5 instead of  3)

Can you confirm this? Thank you very much.

**Reply by Sympatron** [reply]
This number depends on the hardware you are using. [See this](https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-definition). It is the rx gpio pin number.

**Reply by Smitty367** [reply]
Oh, yes I understand. This one I wasn't aware of. Thank you very much for that hint . 
kind regards...

**Reply by anro05** [reply]
@FaserF I try to use the script for my SGM-C8 as well. As you can see in topic #22251 I have some more problems with the delivered data which are hopping from the real values to wrong ones like 999.999 kWh. It seems that the IR reader gets new values about every 3 seconds. Do you have any idea how to fix this? Many thanks.

**Reply by FaserF** [reply]
@anro05 I have now posted the script here: https://github.com/arendst/Tasmota/discussions/21415#discussioncomment-10894454

**Reply by Smitty367** [reply]
@FaserF
Thank you very much for your aid Fabian. Do you also know the script lines to get the historical values like consumption yesterday, - last week, - last month, - last year? Would be nice if you could post it too. Many Thanks.

**Reply by FaserF** [reply]
I don't have those sensors. I am just using the script mentioned above. Never needed those values through, as the total sensor has the history in Homeassistent

**Reply by Smitty367** [reply]
@FaserF
You don't need to have special sonsors for historical vlues. Every Smart Meter should give hiszorical data, as for example my Logarex LK13B (see attachment
![historical](https://github.com/user-attachments/assets/b17ca2d3-2821-49c6-82af-c97451f1ac58)
)

**Reply by Smitty367** [reply]
here my Logarex Script. You only have to ask the right variables. But I red that sometimes the manufacturer of the Smart Meter supresses this.

>D
>B
=>sensor53 r
>M 1
+1,3,o,0,9600
1,1-0:16.7.0*255(@1,aktuell,W,current,16
1,1-0:1.8.0*255(@1,gesamt,KWh,total_in,2
1,1-0:1.8.0*96(@1,gestern,KWh,total_day,1
1,1-0:1.8.0*97(@1,letzte Woche,KWh,total_week,1
1,1-0:1.8.0*98(@1,letzter Monat,KWh,total_month,1
1,1-0:1.8.0*99(@1,letztes Jahr,KWh,total_year,1
#

**Comment by gerontec** [comment]
#!/usr/bin/python3
import serial
import time
import binascii
import struct

PORT = '/dev/ttyUSB20'
BAUDRATE = 9600
TIMEOUT = 10  # Increased timeout

def read_sgm_c8():
    try:
        with serial.Serial(PORT, BAUDRATE, timeout=TIMEOUT) as ser:
            print(f"Connected to {PORT} at {BAUDRATE} baud")
            while True:
                data = ser.read(2000)  # Increased buffer size
                if data:
                    print(f"Received {len(data)} bytes")
                    print(f"Raw data: {binascii.hexlify(data).decode()}")
                    if b'\x1b\x1b\x1b\x1b' in data:
                        print("Found start sequence")
                        process_sml_message(data)
                    else:
                        print("Start sequence not found in data")
                else:
                    print("No data received")
                
                time.sleep(1)  # Wait before next read attempt
    except serial.SerialException as e:
        print(f"Serial port error: {e}")
    except KeyboardInterrupt:
        print("Script terminated by user")
    except Exception as e:
        print(f"An error occurred: {e}")

def process_sml_message(data):
    consumption = find_value(data, '77070100010800ff', 3, 10000)
    production = find_value(data, '77070100020800ff', 3, 10000)
    power = find_value(data, '77070100100700ff', 2, 1)
    voltage = find_value(data, '77070100200700ff', 2, 10)
    current = find_value(data, '770701001f0700ff', 2, 1000)

    print(f"Total Consumption: {consumption:.4f} kWh")
    print(f"Total Production: {production:.4f} kWh")
    print(f"Current Power: {power:.1f} W")
    print(f"Voltage: {voltage:.1f} V")
    print(f"Current: {current:.3f} A")

    if power is not None and voltage is not None and current is not None and voltage > 0 and current > 0:
        apparent_power = voltage * current
        if apparent_power > 0:
            power_factor = power / apparent_power
            print(f"Power Factor: {power_factor:.3f}")
    
    print("---")

def find_value(data, obis_code, length, scale_factor):
    index = data.find(bytes.fromhex(obis_code))
    if index != -1:
        value_bytes = data[index + len(bytes.fromhex(obis_code)) + 13:index + len(bytes.fromhex(obis_code)) + 13 + length]
        if len(value_bytes) == length:
            return int.from_bytes(value_bytes, 'big') / scale_factor
    return None

if __name__ == "__main__":
    read_sgm_c8()

**Comment by Smitty367** [comment]
@FaserF

Hello Fabian,

you tested the above sml script on a EFR SGM-C8? Did you change anythng with this script? I tried to run it on my tasmota ir-read head as well and and it did not. My Editor even disables the ´activate script´ flag when I try to save. Would you please post your whole script here again? Maybe with a screen shot?

Thank you very much

**Reply by FaserF** [reply]
I have not really changed anything, worked fine so far for me. Running Tasmota 14.1.0

```
>D
>B
->sensor53 r
>M 1
+1,3,s,16,9600,ENERGY
1,77070100010800ff@1000,Comsumption,kWh,Total,4
;for meters measuring feed-in SGM-Cx-xxx2xx
;1,77070100020800ff@1000,Supply,kWh,Supply,4
;next 4 lines for double-tariff meters SGM-Cx-xxxxTx
;1,77070100010801ff@1000,Comsumption_t1,kWh,Total_t1,4
;1,77070100010802ff@1000,Comsumption_t2,kWh,Total_t2,4
;1,77070100020801ff@1000,Supply_t1,kWh,Supply_t1,4
;1,77070100020802ff@1000,Supply_t2,kWh,Supply_t2,4
;all commented lines from here on for 3-phase meter SGM-C4
1,77070100100700ff@1,Actual Power,W,Power,0
1,77070100200700ff@1,Voltage L1,V,Voltage,1
1,77070100340700ff@1,Voltage L2,V,Voltage_L2,1
1,77070100480700ff@1,Voltage L3,V,Voltage_L3,1
1,770701001f0700ff@1,Current L1,A,Current,2
1,77070100330700ff@1,Current L2,A,Current_L2,2
1,77070100470700ff@1,Current L3,A,Current_L3,2
1,77070100510701ff@1,Phaseangle L2-L1,deg,phase_angle_L2_L1,0
1,77070100510702ff@1,Phaseangle L3-L1,deg,phase_angle_L3_L1,0
1,77070100510704ff@1,Phaseangle I/U L1,deg,phase_angle_L1,1 
1,7707010051070fff@1,Phaseangle I/U L2,deg,phase_angle_L2,1  
1,7707010051071aff@1,Phaseangle I/U L3,deg,phase_angle_L3,1 
1,770701000e0700ff@1,Frequency,Hz,Freq,1
;all commented lines from here on just for completeness
;1,7707010000020000@1,Firmware Version,,FwVer,0
;1,77070100605a0201@1,Firmware Checksum,,FwCheck,0
;1,7707010061610000@1,Error Register,,ErrorReg,0
;1,7707010060320101@1,Hardware Version,,HwVer,0
;1,7707010060320104@1,Parameter Version,,ParamVer,0
;1,77070100600100ff@#,Server-ID,,ID,0
;You can find your server-id printed on your meter. If you want you can also convert it to your Identifikationsnummer with some ASCII, DEC and HEX encoding. 0A-01-45-46-52-ST-UV-WX-YZ = 1EFR + string(S) + string(T) + hexToDec(UVWXYZ)
```

**Reply by simoneggersdorfer** [reply]
> I have not really changed anything, worked fine so far for me. Running Tasmota 14.1.0
> 
> ```
> >D
> >B
> ->sensor53 r
> >M 1
> +1,3,s,16,9600,ENERGY
> 1,77070100010800ff@1000,Comsumption,kWh,Total,4
> ;for meters measuring feed-in SGM-Cx-xxx2xx
> ;1,77070100020800ff@1000,Supply,kWh,Supply,4
> ;next 4 lines for double-tariff meters SGM-Cx-xxxxTx
> ;1,77070100010801ff@1000,Comsumption_t1,kWh,Total_t1,4
> ;1,77070100010802ff@1000,Comsumption_t2,kWh,Total_t2,4
> ;1,77070100020801ff@1000,Supply_t1,kWh,Supply_t1,4
> ;1,77070100020802ff@1000,Supply_t2,kWh,Supply_t2,4
> ;all commented lines from here on for 3-phase meter SGM-C4
> 1,77070100100700ff@1,Actual Power,W,Power,0
> 1,77070100200700ff@1,Voltage L1,V,Voltage,1
> 1,77070100340700ff@1,Voltage L2,V,Voltage_L2,1
> 1,77070100480700ff@1,Voltage L3,V,Voltage_L3,1
> 1,770701001f0700ff@1,Current L1,A,Current,2
> 1,77070100330700ff@1,Current L2,A,Current_L2,2
> 1,77070100470700ff@1,Current L3,A,Current_L3,2
> 1,77070100510701ff@1,Phaseangle L2-L1,deg,phase_angle_L2_L1,0
> 1,77070100510702ff@1,Phaseangle L3-L1,deg,phase_angle_L3_L1,0
> 1,77070100510704ff@1,Phaseangle I/U L1,deg,phase_angle_L1,1 
> 1,7707010051070fff@1,Phaseangle I/U L2,deg,phase_angle_L2,1  
> 1,7707010051071aff@1,Phaseangle I/U L3,deg,phase_angle_L3,1 
> 1,770701000e0700ff@1,Frequency,Hz,Freq,1
> ;all commented lines from here on just for completeness
> ;1,7707010000020000@1,Firmware Version,,FwVer,0
> ;1,77070100605a0201@1,Firmware Checksum,,FwCheck,0
> ;1,7707010061610000@1,Error Register,,ErrorReg,0
> ;1,7707010060320101@1,Hardware Version,,HwVer,0
> ;1,7707010060320104@1,Parameter Version,,ParamVer,0
> ;1,77070100600100ff@#,Server-ID,,ID,0
> ;You can find your server-id printed on your meter. If you want you can also convert it to your Identifikationsnummer with some ASCII, DEC and HEX encoding. 0A-01-45-46-52-ST-UV-WX-YZ = 1EFR + string(S) + string(T) + hexToDec(UVWXYZ)
> ```

Hey man, I wanted to do the same Stromzähler Model i. home Assistant. Could you maybe get in contact with me and help me witj the full setup and hardware? Would be really happy!!

**Reply by YoungJules** [reply]
Hi. I have an SGM-C8-A620S and am interested to share details on how to (for e.g.) convert the Meter ID to the EFR number that's printed on the meter. Can it be done in the script itself?
Also I'm running Tasmota 13.2.0 and would like to know where you're downloading later versions that still include the possibility to add scripting... Many thanks!
For the rest, it's working nicely in HA, can see power usage etc. via MQTT...

**Reply by simoneggersdorfer** [reply]
> Hi. I have an SGM-C8-A620S and am interested to share details on how to (for e.g.) convert the Meter ID to the EFR number that's printed on the meter. Can it be done in the script itself? Also I'm running Tasmota 13.2.0 and would like to know where you're downloading later versions that still include the possibility to add scripting... Many thanks! For the rest, it's working nicely in HA, can see power usage etc. via MQTT...

Hey, du bist glaube ich auch aus Deutschland oder? Kannst du mich vielleicht privat irgendwie kontaktieren? Ich habe das Modell und würde das Ganze gerne endlich umsetzen! :)

**Reply by YoungJules** [reply]
Hey, I think my LinkedIn is (now) on my profile...

**Comment by YoungJules** [comment]
Just FYI, I managed to update my Tasmota to the latest. I downloaded the development version and compiled it myself in VS Code. Pretty smooth, it even sees it needs PlatformIO and installs it for you.
Still trying to figure out a way to read hardware version and the meter id...

**Comment by YoungJules** [comment]
OK the firmware version is ASCII not a float, so (from the raw dump of the SML data) it appears as:
`30 31 2e 30 33`
which is version 01.03 ... anyone have any ideas if you can show ASCII in Tasmota script? 
The meter id in hex in the SML data dump is (last 8 hex digits masked with x):
`45 46 52 22 xx xx xx xx`
So the 45 46 52 22 seems to relate to the first part of the meter id printed on the meter i.e. EFR22. It's sort of ASCII but not quite, as the 22 is the actual hex value after the normal ASCII representation of the string EFR.
The next 8 hex digits when converted to decimal do correspond with the last part (8 digits in the form nnnn nnnn) of the meter id printed on the meter.
Not sure how paranoid I have to be so have masked the actual meter id just in case.
